### DISCUSSION TITLE
TTGO T-Call + SD-Card

### DISCUSSION THREAD

**Discussion by tadejales** [micropython/micropython#15263] [ESP32]
Hello,

I have a specific problem with my system configuration. I have a TTGO T-Call module with an integrated SIM800L module. I want to add an external SD card module (3.3V) to the system as I want to store data there long-term. The expected size is in the area of hundreds of MB, so I can't just store it in flash memory.

What I want to achieve is to send data over GPRS to the cloud via HTTP POST requests and store the data on the SD card as well for archiving in case transmission fails. Initially, I was just using flash memory to store data as I thought storing them on an SD card later would be trivial. Everything worked fine. Then I added an external SD card, and now I have a problem with transmission. If I initialize the SIM module first, I am able to obtain an IP address and connection, but later I cannot transmit data. It fails even to get the signal strength.

Then I tried initializing the SD card first and the SIM module after that, and the SIM module initialization fails. Based on this, I assume I am having a problem with power. I am using an external power supply of 24V@1A and then stepping down the voltage with a voltage converter:
![image](https://github.com/micropython/micropython/assets/22613409/69e97b83-9c69-49c4-a51a-5400daf887db)

Is there a way to solve the issue without changing the power supply, perhaps by adding some inductors and/or capacitors to stabilize the circuit?

**Comment by davefes** [comment]
That buck converter is rated at 3A, which you should be able to get from your 24V 1A supply.  Actually, with 100uH about 2.5A. I would be probing the 5V with an oscilloscope before I changed the power setup.  If you see voltage dips maybe try a few more 1000uF low-ESR caps on the output.

**Reply by tadejales** [reply]
Any idea what to do if I do not have access to the oscilloscope. I have a multimeter and that is it.

So you suggest something like this:
![image](https://github.com/micropython/micropython/assets/22613409/de3d0de0-e2e1-41e3-9337-a58bba602099)

**Reply by davefes** [reply]
Correct, but I doubt this will fix your problem. Can your device operate off a single Li-ion cell?

**Reply by tadejales** [reply]
No, it has to be wall powered as it will monitor the situation constantly, for months or even years at a time, so battery operation is not an option.

**Reply by davefes** [reply]
But, to rule-out a power supply issue a short-term test with a Li-ion cell that can deliver high peak-current might help.

Is your multimeter a DVM? I have seen mine respond to quick changes in supply voltages.

**Reply by davefes** [reply]
If both the SD card and SIM module are on the same SPU bus there have been some recent discussions here.

**Reply by tadejales** [reply]
Adding more capacitors did not solve the issue as you have already predicted. 

> But, to rule-out a power supply issue a short-term test with a Li-ion cell that can deliver high peak-current might help.
> 
> Is your multimeter a DVM? I have seen mine respond to quick changes in supply voltages.

I dont even have access to the Li-ion cell. Will try to get access to one.



> If both the SD card and SIM module are on the same SPU bus there have been some recent discussions here.

What do you mean by this?

**Reply by davefes** [reply]
> What do you mean by this?

Sorry, that should be SPI bus.

https://github.com/orgs/micropython/discussions/15214

What firmware are you trying to run on this device?  I bought one of these devices about 2 years ago to find out 2G was going to disappear in NZ.

**Reply by tadejales** [reply]
I am not really experienced with SPI interface. If I am honest I do not really understand the background and what is meant there. If I am not mistaken they run on different pins, at least some of them, but I dont know if that means different SPI bus. 

```
# Pin SD SPI
MOSI_PIN = 23
MISO_PIN = 19
SCK_PIN = 18
CS_PIN = 5

# SIM800L
MODEM_TX_PIN = 26
MODEM_RX_PIN = 27
MODEM_PWKEY_PIN = 4
MODEM_RST_PIN = 5
MODEM_POWER_ON_PIN = 23
```

For the firmware I am using the following -> ESP32_GENERIC-SPIRAM-20240222-v1.22.2

**Reply by davefes** [reply]
Looks like SPI for the SD card and UART for the SIM800L.  Before getting a Li-ion cell you need to be certain that the TTGO T-Call will still operate down around 4Volts.

https://lastminuteengineers.com/sim800l-gsm-module-arduino-tutorial/

Could you link to the code that you are putting on this device?  See you tomorrow.

**Reply by tadejales** [reply]
I can share some of it, but not the whole codebase, it is too large and too complex at this point. I did my best and tried to retain just the relavant parts. 

[codebase.zip](https://github.com/user-attachments/files/15804681/codebase.zip)

Regarding the voltage, the sim800L should operate just fine at that voltage. Will see what you find in the codebase.

**Reply by davefes** [reply]
The  code ... is way beyond me.  I have had one attempt at asyncio and it looks like you might be using `workers-framework-micropython`, which I have only read about.

Could you run a test program that just initialises the radio, then the SD card and then blinks an LED?

How long is 2G going to be supported in your country?

**Reply by davefes** [reply]
Where do you `import sdcard`?

**Reply by davefes** [reply]
BTW, the best $60NZ I have spent: `ZEEWEI DSO154Pro` found on their store at Aliexpress.

**Reply by davefes** [reply]
The other discussion that may help:
https://github.com/orgs/micropython/discussions/14502

**Reply by tadejales** [reply]
> The code ... is way beyond me. I have had one attempt at asyncio and it looks like you might be using `workers-framework-micropython`, which I have only read about.
> 
> Could you run a test program that just initialises the radio, then the SD card and then blinks an LED?
> 
> How long is 2G going to be supported in your country?

I am not using  `workers-framework-micropython`, Just asyncio and tasks. 

I will create MVP soon, then go from there. 

2G is predicted to be shut down at 2030.

> Where do you `import sdcard`?

Import and init is done inside `database_handler.py`

**Reply by davefes** [reply]
Maybe, try SoftSPI and see if that changes behaviour.

Good luck!

**Reply by tadejales** [reply]
So I have taken my time to create MVP model with GSM and SD card module to test the functionality. 

```
from lib.sim800l import Modem, Response
import sdcard 
from machine import  SPI, Pin
import utime as time
import json
import os

class Gsm():

    def __init__(self) -> None:
        self.config = {"username": "", "password": "", "apn": "internet.t-2.net"}

        self.modem = Modem(MODEM_POWER_ON_PIN=23, MODEM_PWKEY_PIN=4, MODEM_RST_PIN=5, MODEM_TX_PIN=26, MODEM_RX_PIN=27) 
        # Initialize the modem
        self.modem.initialize()

        timeout = 10
        try:
            self.modem.connect(apn=self.config["apn"], user=self.config["username"], pwd=self.config["password"])
            for _ in range(timeout):
                if self.modem.get_ip_addr():
                    print(f"GPRS connected: {self.modem.get_ip_addr()}")
                    break
                elif _ == timeout-1:
                    print(f"TIMEOUT: Unable to obtain GPRS connection")
                else:
                    time.sleep(1)

        except Exception as e:
            print(f"Error during GPRS connection establishment: {e}")

    def http_request(self, url:str, mode: str='GET', data: any=None, content_type: str='application/json', auth_token: str=None) -> Response:
        return self.modem.http_request(url=url, mode=mode, data=data, content_type=content_type, auth_token=auth_token)

def sd_init():
    print("Trying to mount SD card")
    spi = SPI(1, sck=Pin(18), mosi=Pin(23), miso=Pin(19))
    db = sdcard.SDCard(spi, Pin(5))
    os.mount(db, "/db")

def test_write_read():

    # Test write
    print("Testing write")
    with open("/db/test.txt", "w") as f:
        f.write("TEST")

    # Test read
    print("Testing read")
    with open("/db/test.txt", "r") as f:
        print(f"File contents: {f.read()}")

def boot():
    # Start modem
    gsm = Gsm()

    # Test SD card interaction
    sd_init()
    test_write_read()

    # Verify if modem connection still live
    print("Verifying modem connection")
    if gsm.modem.get_ip_addr():
        print(f"Modem connected with IP: {gsm.modem.get_ip_addr()}")
    else:
        print("Modem not connected")

# RUN
boot()
```

So here are my findings:
1. If I comment out `sd_init()` and `test_write_read()` in `boot()` modem connects and verifys the connection
```
GPRS connected: 10.20.186.123
Verifying modem connection
Modem connected with IP: 10.20.186.123
```
2. If I init GSM module first it establishes the connection. After initing SD card and write/read the modem loses connection. Output bellow:
```
GPRS connected: 10.20.201.231
Trying to mount SD card
Testing write
Testing read
File contents: TEST
Verifying modem connection
Traceback (most recent call last):
  File "boot.py", line 68, in <module>
  File "boot.py", line 62, in boot
  File "lib/sim800l.py", line 287, in get_ip_addr
  File "lib/sim800l.py", line 183, in execute_at_command
Exception: Timeout for command "getbear" (timeout=3)
```
3. If I put `sd_test()` and `test_write_read()` above `gsm = Gsm()` and init SD card before GSM module, gsm gets connected afterwards. Output bellow:
```
Trying to mount SD card
Testing write
Testing read
File contents: TEST
GPRS connected: 10.40.216.190
Verifying modem connection
Modem connected with IP: 10.40.216.190
```
4. If I put `sd_test()` and `test_write_read()` above `gsm = Gsm()` and init SD card before GSM module, gsm gets connected afterwards. If I then add `test_write_read()` after IP check SD card interface fails. Output bellow:
```
Trying to mount SD card
Testing write
Testing read
File contents: TEST
GPRS connected: 10.20.226.10
Verifying modem connection
Modem connected with IP: 10.20.226.10
Testing write
Traceback (most recent call last):
  File "boot.py", line 79, in <module>
  File "boot.py", line 68, in boot
  File "boot.py", line 45, in test_write_read
  File "/lib/sdcard.py", line 238, in readblocks
OSError: [Errno 5] EIO
```

I hope you can take a bit of time if you see any concerning parts here or give me any suggestion what to do. I will take my time ang go thrue the threads that you have sugested looking at. 

From what I see the problem might be in pins 5 and 23 as they are used in both instances.

**Reply by davefes** [reply]
There are a few SPI experts around.  I didn't pickup on the duplication of pins.  I think for SPI you need at least a separate reset pin for each device.  If you have some spare GPIO I would try a SoftSPI on them. 

Edit:
In `codebase.zip` 
```py
# Pin SD SPI
MOSI_PIN = const(23)
MISO_PIN = const(19)
SCK_PIN = const(18)
CS_PIN = const(5)

# SIM800L
MODEM_TX_PIN = const(26)
MODEM_RX_PIN = const(27)
MODEM_PWKEY_PIN = const(4)
MODEM_RST_PIN = const(5)
MODEM_POWER_ON_PIN = const(23)
`
maybe it is that pin 23.  I see where I got confused ... the names for the SIM800L pins suggested that it was driven via a UART, not SPI.

**Reply by robert-hh** [reply]
> a separate reset pin for each device

a separate chip select pin for each device. And you may have to adapt the baud rates for each device.

**Reply by davefes** [reply]
Be aware that there are only 2 hardware SPI buses on specific pins.
https://docs.micropython.org/en/latest/esp32/quickref.html

**Comment by jonnor** [comment]
Have you checked the data lines to/from the the SIM800 with an oscilloscope - ensuring that the signal integrity is high, both generally and not too much affected by SD card communication? Also, be sure to verify the timings on the serial connection - if they are borderline, corruptions can trigger from slight temperature/humidty/voltage changes.

SIM800L, and GSM generally, is a power hungry beast. Major PITA compared to 4G LTE modems. Would not recommend for new design.

**Reply by tadejales** [reply]
I do not have oscilloscope neither do I have the knowlage about the things you are explaining to me. The problem was that some pins were shared. Now I am using different pins for GSM and SD card and it seems to work fine. 

Another question would be; Are LTE modules easier to use and simpler to implement, hardware and software wise? I can still alter the design. I am just using TTGO T-call v1.3 because I got it from the friend and it was the only one that had everything onboard. Now I was looking at the TTGO site and v1.4 has option with LTE module. Would you recommed using that?
